### PR TITLE
board: arm: add support for waveshare_open103z

### PR DESCRIPTION
**Pull request by smanni**
Add support for the board Waveshare Open103Z
based on a STM32F103Zx processor.

Signed-off-by: Stefano Manni <stefano.manni@gmail.com>

### CODE REVIEW

File: boards/arm/waveshare_open103z/waveshare_open103z.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  - gnuarmemb
+  - xtools
+ram: 64
+flash: 512
+supported:
+  - gpio
+  - spi
+  - pwm
+  - watchdog
+  - adc
```
Comment (erwango) [inline boards/arm/waveshare_open103z/waveshare_open103z.yaml:16]: usb_device

File: boards/arm/waveshare_open103z/waveshare_open103z_defconfig [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+# enable pinmux
+CONFIG_PINMUX=y
+
+# enable GPIO
+CONFIG_GPIO=y
+
+# enable SPI
+CONFIG_SPI=y
+CONFIG_SPI_STM32=y
+CONFIG_SPI_STM32_USE_HW_SS=y
```
Comment (erwango) [inline boards/arm/waveshare_open103z/waveshare_open103z_defconfig:23]: Peripheral activation (other than basics gpio, serial) is not part of default board configuration, this is application responsibility.
Also, the Kconfig system will take care of activating SPI_STM32 based on SPI
These should be removed.

File: boards/arm/waveshare_open103z/waveshare_open103z_defconfig [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+# enable I2C
+CONFIG_I2C=y
```
Comment (erwango) [inline boards/arm/waveshare_open103z/waveshare_open103z_defconfig:26]: Ditto

File: boards/arm/waveshare_open103z/waveshare_open103z_defconfig [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+CONFIG_I2C_STM32=y
+
+# enable CAN
+CONFIG_CAN=y
```
Comment (erwango) [inline boards/arm/waveshare_open103z/waveshare_open103z_defconfig:30]: Ditto

File: boards/arm/waveshare_open103z/waveshare_open103z_defconfig [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+CONFIG_CAN_STM32=y
+
+# enable USB
+CONFIG_USB_DC_STM32=y
```
Comment (erwango) [inline boards/arm/waveshare_open103z/waveshare_open103z_defconfig:34]: Ditto

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this addition.
Minor tweaks, otherwise this  looks good

**Comment by erwango** [comment]
@smanni please squash your commits and fix `WARNING:BLOCK_COMMENT_STYLE` reported by zephyrbot

**Comment by smanni** [comment]
@erwango please check now

**Comment by smanni** [comment]
@erwango does it need further fixes?

File: boards/arm/waveshare_open103z/pinmux.c [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+#include <device.h>
+#include <init.h>
+#include <drivers/pinmux.h>
+#include <sys/sys_io.h>
+
+#include <pinmux/stm32/pinmux_stm32.h>
+
+/* pin assignments for waveshre_open103z board */
+static const struct pin_config pinconf[] = {
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(usart1), okay)
```
Comment (erwango) [inline boards/arm/waveshare_open103z/pinmux.c:17]: On master, we're back adding Kconfig control here:
`#if DT_NODE_HAS_STATUS(DT_NODELABEL(usart1), okay) && CONFIG_SERIAL`
Please update for all devices

**Review by erwango — CHANGES_REQUESTED** [review]
Some change required due to latest evolution on master

File: boards/arm/waveshare_open103z/pinmux.c (hunk 6 above)
Comment (smanni) [reply boards/arm/waveshare_open103z/pinmux.c]: just force pushed with the required changes

**Review by erwango — APPROVED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
@smanni Can you rebase and try to fix error reported by shippable ?

**Comment by smanni** [comment]
> @smanni Can you rebase and try to fix error reported by shippable ?

@erwango all checks are passing now. Can it be merged into master?

**Review by erwango — CHANGES_REQUESTED** [review]
This is globally fine to me but please split the adc test change into a separate commit.

**Comment by smanni** [comment]
> This is globally fine to me but please split the adc test change into a separate commit.

done.

**Comment by smanni** [comment]
@erwango I'm wondering if this PR will be merged. I have some branches waiting to be submitted that rely on having this board in tree.

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
Sanity issue: `Sanitycheck / qemu_arc_em:tests/kernel/lifo/lifo_usage/kernel.lifo.usage / tests/kernel/lifo/lifo_usage/kernel.lifo.usage`

File: boards/arm/waveshare_open103z/waveshare_open103z.dts [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+	phase-seg2 = <6>;
+};
+
+&usb {
+	/*
+	 * make sure CAN and USB are not enabled at the same time
+	 * because they share interrupts 19, 20 (stm32f103Xb.dtsi)
+	 * reference: RM0008 rev20 page 205
+	 */
+	status = "okay";
```
Comment (FRASTM) [inline boards/arm/waveshare_open103z/waveshare_open103z.dts:127]: in the DTS, stm32f103Xe which includes stm32f103Xb, has USB interrupt <20>
( interrupt <19> is for _USB high priority or CAN TX_ interrupt on  XL-density devices )

Comment (smanni) [reply boards/arm/waveshare_open103z/waveshare_open103z.dts:127]: In stm32f103Xb.dtsi I see that USB uses interrupt 20.

Here is an extract from RM008 Table 63:
RM008-table63

How to proceed? Shall I remove that comment?

Comment (FRASTM) [reply boards/arm/waveshare_open103z/waveshare_open103z.dts:127]: no need, that's fine as is.

**Review by FRASTM — APPROVED** [review]

**Comment by smanni** [comment]
@nashif why did you cherry pick just one of the two commits?

**Pull request merged**